Structure (Protobuf) - Aware Fuzzing Support for Apollo Unit Test

jiayunhan opened the pull request:
Add Protobuf-aware fuzzing support. Made the following changes:
- Add a external git repository [libprotobuf-mutator for Apollo](https://github.com/jiayunhan/libprotobuf-mutator) into WORKSPACE.in, which is a stable version that provides the mutation support for Protobuf messages. 
- Add lat_controller_fuzzer.cc, lon_controller_fuzzer.cc, simple_control_fuzz.cc, and some other fuzzing testcases as examples, together with their BUILD files. 
- Add a CROSSTOOL for clang-6.0, which is required to use libFuzzer. 
- Modified apollo.sh to exclude fuzzer unit test from the build command, since they rely on CROSSTOOL. The fuzzer unit test can only be built using bazel command manually. 

PASS all CPP Lint check.

Detailed introduction, and instructions for this pull request, can be found in this [README](https://github.com/jiayunhan/libprotobuf-mutator). 
If you have any questions, ping me on Baidu Hi. Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ApolloAuto/apollo?pullRequest=4945) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ApolloAuto/apollo?pullRequest=4945) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ApolloAuto/apollo?pullRequest=4945) it.</sub>

xiaoxq commented on WORKSPACE.in:
```diff
@@ -203,3 +203,10 @@ new_local_repository(
     build_file = "third_party/adv_plat.BUILD",
     path = "/usr/local/apollo/adv_plat"
 )
+#libprotobuf-mutator
```
Blank line before the section and new line after.

xiaoxq commented on WORKSPACE.in:
```diff
@@ -203,3 +203,10 @@ new_local_repository(
+    name = "libprotobuf_mutator",
+    build_file = "third_party/libprotobuf_mutator.BUILD",
+    commit = "91a872998bc204e9484dccdbeeafc608c1ee8062",
+    remote = "https://github.com/jiayunhan/libprotobuf-mutator",
```
Better to refer an official repo instead of personal.

jiayunhan replied:
I forked this repo from Google Github with some modifications, where do you suggest to put the repo? Is it possible to put it together with Apollo, under ApolloAuto. I will fix other issues. Thanks.

jiayunhan replied:
Never mind, I will put it under our  Baidu X-Lab Github organization. Thanks.

xiaoxq commented on modules/control/control.h:
```diff
@@ -98,7 +98,7 @@ class Control : public apollo::common::ApolloApp {
   void SendCmd(ControlCommand *control_command);
 
- private:
+  private:
```
Should not change. Please run "apollo.sh lint" and fix all style issues.

xiaoxq commented on modules/control/controller/BUILD:
```diff
@@ -189,5 +189,40 @@ cc_test(
     ],
 )
-
+cc_binary(
```
Blank line before the section and new line after.

xiaoxq commented on modules/control/controller/BUILD:
```diff
@@ -189,5 +189,40 @@ cc_test(
+        "//modules/planning/proto:planning_proto",
+        "@libprotobuf_mutator//:mutator",
+    ],
+    copts = ["-fsanitize=fuzzer,address",
```
List in rows like the "deps" section.

xiaoxq commented on modules/control/controller/BUILD:
```diff
@@ -189,5 +189,40 @@ cc_test(
+        "//modules/planning/proto:planning_proto",
+        "@libprotobuf_mutator//:mutator",   
+    ],
+    copts = ["-fsanitize=fuzzer,address",
```
ditto

xiaoxq commented on modules/control/controller/lat_controller_fuzzer.cc:
```diff
@@ -0,0 +1,147 @@
+/******************************************************************************
+ * Copyright 2017 The Apollo Authors. All Rights Reserved.
+ * Copyright 2018 Baidu X-Lab. Yunhan Jia <jiayunhan@baidu.com>
```
We would prefer unified copyright in the master repo.

jiayunhan commented:
Made the required changes. Passed cpp lint, fixed copyright statements. Excludes all fuzz target from apollo.sh build, since they rely on clang. Currently they can only be built manually using bazel command. Moved the third-party repository to Baidu X-Lab organization.
